onyo/lib/items.py

```diff
@@ -0,0 +1,167 @@
+from __future__ import annotations
+
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import onyo.lib.onyo
+import onyo.lib.inventory
+import onyo.lib.pseudokeys
+from onyo.lib.utils import DotNotationWrapper
+
+
+if TYPE_CHECKING:
+    from typing import (
+        Any,
+        Mapping,
+        TypeVar,
+    )
+
+    _KT = TypeVar("_KT")  # Key type.
+    _VT = TypeVar("_VT")  # Value type.
+
+
+def resolve_alias(key: Any) -> Any:
+    """Return the target of a key alias."""
+    try:
+        return onyo.lib.pseudokeys.PSEUDOKEY_ALIASES[key]
+    except KeyError:
+        return key
+
+
+class Item(DotNotationWrapper):
+    """An Item an Inventory can potentially track.
+
+    The main purpose of this class is to attach pseudo-keys
+    and alias resolution to things that can be inventoried.
+    That's directories and YAML-files.
+
+    The initializer methods are referenced in the `PSEUDO_KEYS`
+    mapping. They are called, when a pseudo-key is first
+    accessed (see `self.__getitem__`). This allows to distinguish
+    a meaningful `None` (<unset>) from a not yet evaluated
+    pseudo-key.
+    """
+
+    def __init__(self,
+                 item: Mapping[_KT, _VT] | Path | None = None,
+                 repo: onyo.lib.onyo.OnyoRepo | None = None,
+                 **kwargs: _VT):
+        super().__init__()
+        self.repo: onyo.lib.onyo.OnyoRepo | None = repo
+        self.update(onyo.lib.pseudokeys.PSEUDO_KEYS)
+        self._path = None
+
+        if isinstance(item, Path):
+            assert item.is_absolute()  # currently no support for relative. This is how all existing code should work ATM.
+            self.update_from_path(item)
+        elif item is not None:
+            self.update(item)
+
+        if kwargs:
+            self.update(**kwargs)
+
+    def __setitem__(self, key, value):
+        key = resolve_alias(key)
+        super().__setitem__(key, value)
+
+    def __getitem__(self, key):
+        key = resolve_alias(key)
+        value = super().__getitem__(key)
+        if key in onyo.lib.pseudokeys.PSEUDO_KEYS and \
```

Comment:
This seems to do a lookup every time, with no caching.

Reply:
Do you mean WRT aliases or the testing for whether this is still `PseudoKey`?
The former: Yes, that'll need to rely on config caching.
The latter: Would be nice to cache, but not yet sure how. That's because I'm not yet sure how to properly do the various ways of stripping/resetting/ignoring pseudo-keys in different operations/commands. These patterns will hopefully lead to proper `Item` methods, that do that. Once that's clear, we know how to cache (and how o invalidate it)

Reply:
> Do you mean WRT aliases or the testing for whether this is still PseudoKey?

Neither. I mean, once we've looked up the pseudo-key value once (such as `onyo.is.directory`), we don't store it, and will execute it again if requested again.

I can understand not doing this now, and waiting until more pieces land to determine if caching such lookups is practical. But yeah, that's what I was pointing out.

Reply:
Have you seen line 75 below? It is stored. Not a cache exactly, but that's probably not the point.

Reply:
The magic for me is in line 71. I understand it now. Thanks!